### PR TITLE
decouple QML from legacy skin system + add preferences button

### PR DESCRIPTION
**Pull request by Be-ing**
Somehow this is hanging in the ControllerManager destructor. It seems the `requestShutdown` signal does not trigger `slotShutdown` so `m_pThread->wait()` blocks indefinitely.

### CODE REVIEW

File: src/qmlapplication.h
Comment (uklotzde) [inline src/qmlapplication.h:28]: We should not store a plain reference to CmdlineArgs as a member and instead obtain it with CmdlineArgs::Instance(). This is already wrong in CoreServices, I could submit a PR to fix that.

File: src/qmlapplication.h
Comment (uklotzde) [inline src/qmlapplication.h:20]: CmdlineArgs is a singleton and does not need to be passed around -> remove

**Review by uklotzde — CHANGES_REQUESTED** [review]

**Comment by Be-ing** [comment]
Any ideas why ~ControllerManager is hanging?

**Comment by uklotzde** [comment]
> Any ideas why ~ControllerManager is hanging?

I didn't have time to test it yet. Will do later.

File: src/qmlapplication.h
Comment (Be-ing) [reply src/qmlapplication.h]: CmdlineArgs was unused copypasta from #3448 anyway. Removed and force pushed.

**Comment by Be-ing** [comment]
Effects including EQs are not working with QML. I tested merging #2618 but that does not fix it. There must be some implicit dependency of the effects system on the setup of the legacy GUI.

**Comment by Be-ing** [comment]
Well this is really ugly. Initialization of the effects system depends on the existence of a DlgPreferences instance.

**Comment by Be-ing** [comment]
Even uglier: the hang in ~ControllerManager goes away when a DlgPreferences instance is created.

File: src/qml/qmlapplication.h
Comment (uklotzde) [inline src/qml/qmlapplication.h:18]: `override`

File: src/qml/qmlapplication.h
Comment (uklotzde) [inline src/qml/qmlapplication.h:26]: `m_rootQmlFilePath`

File: src/qml/qmlapplication.h
Comment (uklotzde) [inline src/qml/qmlapplication.h:29]: Shall we drop the redudndant "qml/""Qml" term from all members? It seems to be redundant. This class should not take more responsibilities so it might to be safe.

**Review by uklotzde — COMMENTED** [review]
Dropping the "skin" prefix from namespaces and file paths is a good idea. We have *skins* and we have *qml*, mutually exclusive. This avoids to distinguish between *legacy skins* and *QML skins*.

File: src/qml/qmlapplication.h
Comment (Be-ing) [reply src/qml/qmlapplication.h:18]: default override? Would it be better to simply remove this line?

File: src/qml/qmlapplication.h
Comment (Be-ing) [reply src/qml/qmlapplication.h:26]: The QML file is named `main.qml` so I think this variable name is appropriate.

File: src/qml/qmlapplication.h
Comment (Be-ing) [reply src/qml/qmlapplication.h:29]: Yes, good idea.

File: src/qml/qmlapplication.h
Comment (uklotzde) [reply src/qml/qmlapplication.h:18]: No, `override` should be used for all destructors of derived classes as a convention. C++ is so error prone and requires a lot of informal boilerplate.

Comment (uklotzde) [reply src/qml/qmlapplication.h:18]: If omitted the base class is not checked for a virtual destructor. QObject has one, but nevertheless follow the convention. If you inherit than add either a `virtual` or an `override` destructor.

**Comment by Be-ing** [comment]
More ugliness: I cannot include include preferences/dialog/dlgpreferences.h in src/qml/qmlapplication.h because compiling main.cpp fails with:
```
In file included from ../src/qml/qmlapplication.h:4,
                 from ../src/main.cpp:14:
../src/preferences/dialog/dlgpreferences.h:13:10: fatal error: preferences/dialog/ui_dlgpreferencesdlg.h: No such file or directory
   13 | #include "preferences/dialog/ui_dlgpreferencesdlg.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

I tried adding

```
add_executable(mixxx WIN32 src/main.cpp src/qml/qmlapplication.cpp)
set_target_properties(mixxx PROPERTIES AUTOUIC ON AUTOMOC ON)
```
to CMakeLists.txt and removing src/qml/qmlapplication.cpp from the sources list for the mixxx-lib target. Unfortunately the build still fails. CMake is generating ui_dlgpreferencesdlg.h but it is located in build/mixxx-lib_autogen/include/preferences/dialog/ui_dlgpreferencesdlg.h which is not in the list of include paths for the mixxx executable target. Any suggestions how to resolve this?

File: src/qml/qmlapplication.h
Comment (Be-ing) [reply src/qml/qmlapplication.h:30]: Should remove the `Qml` prefix from the class names within the `mixxx::qml` namespace?

File: src/qml/qmlapplication.h
Comment (Be-ing) [reply src/qml/qmlapplication.h]: Fixed and squashed into the first commit.

**Comment by Be-ing** [comment]
I made a QML button to show the preferences but I am puzzled how closing the preferences window triggers the QmlApplication destructor. I set a breakpoint in ~DlgPreferences which shows that DlgPreferences is properly getting deleted by QmlApplication... but why is QmlApplication getting destroyed in the first place when DlgPreferences is closed?? It does not matter whether DlgPreferences is closed by the Okay, Apply, or Cancel button or the window manager close button. I commented out adding all the pages in the DlgPreferences constructor but I still reproduced the issue. I also tried creating a QQuickView instead of a QQmlApplicationEngine but the issue still occurred.

```
0  DlgPreferences::~DlgPreferences() (this=0x9c1d060, __in_chrg=<optimized out>) at ../src/preferences/dialog/dlgpreferences.cpp:297
#1  0x000000000059394a in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() (this=0x9c1d050)
    at /usr/include/c++/11/bits/shared_ptr_base.h:168
#2  std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() (this=0x9c1d050) at /usr/include/c++/11/bits/shared_ptr_base.h:161
#3  std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count() (this=0x7fffffffdc48, __in_chrg=<optimized out>)
    at /usr/include/c++/11/bits/shared_ptr_base.h:705
#4  std::__shared_ptr<DlgPreferences, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr() (this=0x7fffffffdc40, __in_chrg=<optimized out>)
    at /usr/include/c++/11/bits/shared_ptr_base.h:1154
#5  std::shared_ptr<DlgPreferences>::~shared_ptr() (this=0x7fffffffdc40, __in_chrg=<optimized out>) at /usr/include/c++/11/bits/shared_ptr.h:122
#6  mixxx::qml::QmlApplication::~QmlApplication() (this=0x7fffffffdc00, __in_chrg=<optimized out>) at ../src/qml/qmlapplication.h:19
#7  0x00000000005740c9 in (anonymous namespace)::runMixxx (args=<optimized out>, pApp=0x7fffffffdbc0) at ../src/main.cpp:41
#8  main(int, char**) (argc=<optimized out>, argv=0x7fffffffddd8) at ../src/main.cpp:189
(gdb) bt full
#0  DlgPreferences::~DlgPreferences() (this=0x9c1d060, __in_chrg=<optimized out>) at ../src/preferences/dialog/dlgpreferences.cpp:297
#1  0x000000000059394a in std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() (this=0x9c1d050)
    at /usr/include/c++/11/bits/shared_ptr_base.h:168
#2  std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() (this=0x9c1d050) at /usr/include/c++/11/bits/shared_ptr_base.h:161
#3  std::__shared_count<(__gnu_cxx::_Lock_policy)2>::~__shared_count() (this=0x7fffffffdc48, __in_chrg=<optimized out>)
    at /usr/include/c++/11/bits/shared_ptr_base.h:705
#4  std::__shared_ptr<DlgPreferences, (__gnu_cxx::_Lock_policy)2>::~__shared_ptr() (this=0x7fffffffdc40, __in_chrg=<optimized out>)
    at /usr/include/c++/11/bits/shared_ptr_base.h:1154
#5  std::shared_ptr<DlgPreferences>::~shared_ptr() (this=0x7fffffffdc40, __in_chrg=<optimized out>) at /usr/include/c++/11/bits/shared_ptr.h:122
#6  mixxx::qml::QmlApplication::~QmlApplication() (this=0x7fffffffdc00, __in_chrg=<optimized out>) at ../src/qml/qmlapplication.h:19
#7  0x00000000005740c9 in (anonymous namespace)::runMixxx (args=<optimized out>, pApp=0x7fffffffdbc0) at ../src/main.cpp:41
```

File: src/qml/qmlapplication.cpp
Comment (Be-ing) [inline src/qml/qmlapplication.cpp:76]: I also tried making the DlgPreferences instance a local variable of the QmlApplication constructor and a local variable within this lambda to test if capturing `this` was causing the problem, but neither of those solved the problem.

**Comment by Be-ing** [comment]
Well this is interesting. I replaced the DlgPreferences instance with a simple QDialogButtonBox and reproduced the issue of QmlApplication shutting down when the widget is destroyed.

**Comment by Be-ing** [comment]
I figured out the problem. QApplication's event loop [shuts down the QApplication when the last QWidget window is closed](https://code.woboq.org/qt5/qtbase/src/widgets/kernel/qapplication.cpp.html#2861). I confirmed this by showing a dummy QDialogButtonBox then opening the preferences from QML. In that case, I could close the preferences without shutting down the application but the application shut down after closing both the preferences and the dummy QDialogButtonBox.

I attempted changing MixxxApplication to subclass QGuiApplication instead of QApplication, but then it is not possible to use any QWidgets at all; at runtime console output shows:
```
QWidget: Cannot create a QWidget without QApplication
```

So I think we need to filter out the erroneous QEvent::Close event in MixxxApplication when the QQmlApplicationEngine is still showing to avoid needing to rewrite the whole preferences in QML immediately.

**Comment by Holzhaus** [comment]
Just setting this to false doesn't work?

**Comment by Be-ing** [comment]
Ha! That looks like a simpler solution, thanks.

**Comment by Be-ing** [comment]
Adding
```
    app->setQuitOnLastWindowClosed(false);
    app->setQuitLockEnabled(false);
```
to the QmlApplication constructor does not solve the problem.

**Comment by Be-ing** [comment]
I found a one line fix:
```
m_pDlgPreferences->setAttribute(Qt::WA_QuitOnClose, false);
```
Now to deal with the hang in ~ControllerManager....

**Comment by Be-ing** [comment]
Well this is interesting. I found a workaround for the hack to include the uic generated header for the mixxx executable CMake target. Instead of making DlgPreferences a std::shared_ptr member of QmlApplication, I changed it to a local std::shared_ptr in the QmlApplication constructor. That allowed to move the `#include "preferences/dialog/dlgpreferences.h" to the cpp file which is not `#include`d in the mixxx executable target. Then, magically, ~ControllerManager stopped hanging on shutdown. Go figure.

Now there are more errors on shutdown to address.

**Comment by Be-ing** [comment]
Oof, it seems that DlgPreferences is outliving CoreServices...

**Comment by Be-ing** [comment]
> Then, magically, ~ControllerManager stopped hanging on shutdown.

That is because ~ControllerManager is not getting called because ~CoreServices is running before ~DlgPreferences...

**Comment by Be-ing** [comment]
I think I have figured out how to get the proper lifetime for DlgPreferences and added extensive comments in the QmlApplication constructor explaining this. Now, why is ~ControllerManager hanging...

**Comment by Be-ing** [comment]
The issue with ~ControllerManager hanging seems to be because CoreServices gets destroyed after the Qt event loop stops processing events. That seems to happen because std::shared_ptr references are captured by the lambdas for qmlRegisterSingletonType which outlive QmlApplication.

**Comment by Be-ing** [comment]
Confirmed. The qmlRegisterSingletonType lambdas must not retain ownership of CoreServices or its manager members. Maybe std::weak_ptr can help.

**Comment by Be-ing** [comment]
Yup, the solution was to capture std::weak_ptrs in the lambdas and convert them to std::shared_ptrs when the lambdas are executed by the QQmlEngine to instantiate the QML proxy objects.

**Comment by Be-ing** [comment]
Ready for review.

File: src/errordialoghandler.cpp
Comment (Swiftb0y) [inline src/errordialoghandler.cpp:165]: I'd document why this required.

File: src/preferences/dialog/dlgprefinterface.cpp
Comment (Swiftb0y) [inline src/preferences/dialog/dlgprefinterface.cpp:70]: Have you made sure we don't assume that `m_pSkin != nullptr` anywhere else?

File: res/qml/main.qml
Comment (Swiftb0y) [inline res/qml/main.qml:76]: Not sure if it makes sense for the button to be checkable, when nothing happens when unchecking.

File: src/preferences/dialog/dlgprefinterface.cpp
Comment (Be-ing) [reply src/preferences/dialog/dlgprefinterface.cpp:70]: yes

File: res/qml/main.qml
Comment (Be-ing) [reply res/qml/main.qml:76]: I tried to add a Q_PROPERTY to the QmlDlgPreferencesProxy class indicating the state of the dialog but unfortunately there is no notifyChanged callback for QWidget's `visible` property.

File: src/errordialoghandler.cpp
Comment (Be-ing) [reply src/errordialoghandler.cpp:168]: comment added

File: res/qml/main.qml
Comment (Be-ing) [reply res/qml/main.qml]: Simply removing `checkable: true` works well.

**Review by Swiftb0y — COMMENTED** [review]
LGTM, but I'd like to have a second opinion before merging if possible.

File: src/preferences/dialog/dlgpreferences.cpp
Comment (uklotzde) [inline src/preferences/dialog/dlgpreferences.cpp:155]: typo

File: src/qml/qmlapplication.h
Comment (uklotzde) [reply src/qml/qmlapplication.h]: We could decide about the naming later. Let's leave it as is until we are sure.

**Review by uklotzde — APPROVED** [review]

File: src/preferences/dialog/dlgpreferences.cpp
Comment (Be-ing) [reply src/preferences/dialog/dlgpreferences.cpp]: fixed

**Comment by Be-ing** [comment]
Merge?

**Comment by Holzhaus** [comment]
This PR leads to 2 major regressions:

1. Closing any popup closes Mixxx. E.g. if a track fails to load, a popup will open. If you close it, Mixxx shuts down.
2. It's not possible to use the developer tools anymore, which makes QML skin development extremely tedious.

**Comment by Be-ing** [comment]
>  Closing any popup closes Mixxx. E.g. if a track fails to load, a popup will open. If you close it, Mixxx shuts down.

Qt::WA_QuitOnClose needs to be set to false for every QWidget window: https://github.com/mixxxdj/mixxx/pull/4327/files#diff-2768d0d13eda8f9b18b5ce2c75faed57cb48c7f8463612e674455561533234f7R63

**Pull request merged**
